File: sdk/typescript/test/agent.test.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 The Hive Authors

import { describe, it, before, after, beforeEach, afterEach } from "node:test";
import * as assert from "node:assert/strict";
import * as http from "node:http";
import { HiveAgent } from "../src/index";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Make an HTTP request and return the parsed JSON response. */
async function request(
  method: string,
  url: string,
  body?: Record<string, unknown>
): Promise<{ status: number; body: Record<string, unknown> }> {
  return new Promise((resolve, reject) => {
    const parsedUrl = new URL(url);
    const payload = body ? JSON.stringify(body) : undefined;

    const options: http.RequestOptions = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port,
      path: parsedUrl.pathname,
      method,
      headers: payload
        ? {
            "Content-Type": "application/json",
            "Content-Length": Buffer.byteLength(payload),
          }
        : {},
    };

    const req = http.request(options, (res) => {
      const chunks: Buffer[] = [];
      res.on("data", (chunk: Buffer) => chunks.push(chunk));
      res.on("end", () => {
        const raw = Buffer.concat(chunks).toString("utf-8");
        let parsed: Record<string, unknown> = {};
        try {
          parsed = JSON.parse(raw) as Record<string, unknown>;
        } catch {
          parsed = { raw };
        }
        resolve({ status: res.statusCode ?? 0, body: parsed });
      });
      res.on("error", reject);
    });

    req.on("error", reject);
    if (payload) req.write(payload);
    req.end();
  });
}

/** Wait for a server to accept connections. */
async function waitForServer(
  port: number,
  maxRetries = 20
): Promise<void> {
  for (let i = 0; i < maxRetries; i++) {
    try {
      const res = await request("GET", `http://127.0.0.1:${port}/health`);
      if (res.status === 200) return;
    } catch {
      // Server not ready yet.
    }
    await new Promise((r) => setTimeout(r, 50));
  }
  throw new Error(`Server on port ${port} did not start in time`);
}

// ---------------------------------------------------------------------------
// Tests: Capability Registration
// ---------------------------------------------------------------------------

describe("HiveAgent capability registration", () => {
  it("should register a capability handler", () => {
    const agent = new HiveAgent({ callbackPort: 0 });
    agent.capability("test-cap", async (inputs) => ({
      result: inputs.value,
    }));
    // No error means success. We verify it works via HTTP below.
  });

  it("should throw on empty capability name", () => {
    const agent = new HiveAgent({ callbackPort: 0 });
    assert.throws(
      () => agent.capability("", async () => ({})),
      /capability name must not be empty/
    );
  });

  it("should throw on duplicate capability name", () => {
    const agent = new HiveAgent({ callbackPort: 0 });
    agent.capability("dup", async () => ({}));
    assert.throws(
      () => agent.capability("dup", async () => ({})),
      /already registered/
    );
  });
});

// ---------------------------------------------------------------------------
// Tests: Environment Variable Reading
// ---------------------------------------------------------------------------

describe("HiveAgent environment reading", () => {
  const savedEnv: Record<string, string | undefined> = {};

  before(() => {
    // Save and set env vars.
    const vars = [
      "HIVE_CALLBACK_PORT",
      "HIVE_AGENT_ID",
      "HIVE_TEAM_ID",
      "HIVE_SIDECAR_URL",
      "HIVE_WORKSPACE",
    ];
    for (const v of vars) {
      savedEnv[v] = process.env[v];
    }
    process.env.HIVE_CALLBACK_PORT = "9999";
    process.env.HIVE_AGENT_ID = "env-agent";
    process.env.HIVE_TEAM_ID = "env-team";
    process.env.HIVE_SIDECAR_URL = "http://sidecar:9100";
    process.env.HIVE_WORKSPACE = "/tmp/workspace";
  });

  after(() => {
    // Restore env vars.
    for (const [k, v] of Object.entries(savedEnv)) {
      if (v === undefined) {
        delete process.env[k];
      } else {
        process.env[k] = v;
      }
    }
  });

  it("should read agent ID from env", () => {
    const agent = new HiveAgent();
    assert.equal(agent.agentId, "env-agent");
  });

  it("should read team ID from env", () => {
    const agent = new HiveAgent();
    assert.equal(agent.teamId, "env-team");
  });

  it("should read callback port from env", () => {
    const agent = new HiveAgent();
    assert.equal(agent.callbackPort, 9999);
  });

  it("should read sidecar URL from env", () => {
    const agent = new HiveAgent();
    assert.equal(agent.sidecarUrl, "http://sidecar:9100");
  });

  it("should read workspace from env", () => {
    const agent = new HiveAgent();
    assert.equal(agent.workspace, "/tmp/workspace");
  });

  it("should prefer options over env vars", () => {
    const agent = new HiveAgent({
      agentId: "opt-agent",
      teamId: "opt-team",
      callbackPort: 8888,
      sidecarUrl: "http://opt:9100",
      workspace: "/opt/workspace",
    });
    assert.equal(agent.agentId, "opt-agent");
    assert.equal(agent.teamId, "opt-team");
    assert.equal(agent.callbackPort, 8888);
    assert.equal(agent.sidecarUrl, "http://opt:9100");
    assert.equal(agent.workspace, "/opt/workspace");
  });
});

// ---------------------------------------------------------------------------
// Tests: Default Values
// ---------------------------------------------------------------------------

describe("HiveAgent defaults", () => {
  const savedEnv: Record<string, string | undefined> = {};

  before(() => {
    // Clear all HIVE_ env vars so defaults kick in.
    const vars = [
      "HIVE_CALLBACK_PORT",
      "HIVE_AGENT_ID",
      "HIVE_TEAM_ID",
      "HIVE_SIDECAR_URL",
      "HIVE_WORKSPACE",
    ];
    for (const v of vars) {
      savedEnv[v] = process.env[v];
      delete process.env[v];
    }
  });

  after(() => {
    for (const [k, v] of Object.entries(savedEnv)) {
      if (v === undefined) {
        delete process.env[k];
      } else {
        process.env[k] = v;
      }
    }
  });

  it("should default callback port to 9200", () => {
    const agent = new HiveAgent();
    assert.equal(agent.callbackPort, 9200);
  });

  it("should default sidecar URL to http://127.0.0.1:9100", () => {
    const agent = new HiveAgent();
    assert.equal(agent.sidecarUrl, "http://127.0.0.1:9100");
  });

  it("should default agent ID to empty string", () => {
    const agent = new HiveAgent();
    assert.equal(agent.agentId, "");
  });

  it("should default team ID to empty string", () => {
    const agent = new HiveAgent();
    assert.equal(agent.teamId, "");
  });
});

// ---------------------------------------------------------------------------
// Tests: HTTP Server - Health Endpoint
// ---------------------------------------------------------------------------

describe("HiveAgent HTTP server", () => {
  let agent: HiveAgent;
  // Use port 0 via options; we'll find the actual port from the server.
  // Since run() uses the configured port, we pick a high random port.
  const testPort = 19200 + Math.floor(Math.random() * 1000);

  before(async () => {
    agent = new HiveAgent({
      callbackPort: testPort,
      agentId: "test-agent",
      teamId: "test-team",
    });

    agent.capability("greet", async (inputs) => {
      return { message: `Hello, ${inputs.name}!` };
    });

    agent.capability("fail", async () => {
      throw new Error("something went wrong");
    });

    agent.capability("echo", async (inputs) => {
      return { ...inputs };
    });

    agent.run();
    await waitForServer(testPort);
  });

  after(() => {
    agent.stop();
  });

  it("GET /health returns healthy status", async () => {
    const res = await request("GET", `http://127.0.0.1:${testPort}/health`);
    assert.equal(res.status, 200);
    assert.equal(res.body.status, "healthy");
  });

  it("POST /handle/greet returns outputs", async () => {
    const res = await request(
      "POST",
      `http://127.0.0.1:${testPort}/handle/greet`,
      { inputs: { name: "World" } }
    );
    assert.equal(res.status, 200);
    const outputs = res.body.outputs as Record<string, unknown>;
    assert.equal(outputs.message, "Hello, World!");
  });

  it("POST /handle/echo echoes inputs back", async () => {
    const res = await request(
      "POST",
      `http://127.0.0.1:${testPort}/handle/echo`,
      { inputs: { foo: "bar", count: 42 } }
    );
    assert.equal(res.status, 200);
    const outputs = res.body.outputs as Record<string, unknown>;
    assert.equal(outputs.foo, "bar");
    assert.equal(outputs.count, 42);
  });

  it("POST /handle/fail returns error with 500", async () => {
    const res = await request(
      "POST",
      `http://127.0.0.1:${testPort}/handle/fail`,
      { inputs: {} }
    );
    assert.equal(res.status, 500);
    const error = res.body.error as Record<string, unknown>;
    assert.equal(error.code, "CAPABILITY_FAILED");
    assert.equal(error.message, "something went wrong");
  });

  it("POST /handle/nonexistent returns 404", async () => {
    const res = await request(
      "POST",
      `http://127.0.0.1:${testPort}/handle/nonexistent`,
      { inputs: {} }
    );
    assert.equal(res.status, 404);
    const error = res.body.error as Record<string, unknown>;
    assert.equal(error.code, "CAPABILITY_NOT_FOUND");
  });

  it("GET /nonexistent returns 404", async () => {
    const res = await request(
      "GET",
      `http://127.0.0.1:${testPort}/nonexistent`
    );
    assert.equal(res.status, 404);
  });

  it("POST /handle/greet with empty body uses empty inputs", async () => {
    const res = await request(
      "POST",
      `http://127.0.0.1:${testPort}/handle/greet`
    );
    assert.equal(res.status, 200);
    const outputs = res.body.outputs as Record<string, unknown>;
    assert.equal(outputs.message, "Hello, undefined!");
  });

  it("POST /handle/greet with invalid JSON returns 400", async () => {
    // Send raw invalid JSON.
    const res = await new Promise<{
      status: number;
      body: Record<string, unknown>;
    }>((resolve, reject) => {
      const req = http.request(
        {
          hostname: "127.0.0.1",
          port: testPort,
          path: "/handle/greet",
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "Content-Length": 11,
          },
        },
        (httpRes) => {
          const chunks: Buffer[] = [];
          httpRes.on("data", (c: Buffer) => chunks.push(c));
          httpRes.on("end", () => {
            resolve({
              status: httpRes.statusCode ?? 0,
              body: JSON.parse(
                Buffer.concat(chunks).toString("utf-8")
              ) as Record<string, unknown>,
            });
          });
          httpRes.on("error", reject);
        }
      );
      req.on("error", reject);
      req.write("not valid!!");
      req.end();
    });

    assert.equal(res.status, 400);
    const error = res.body.error as Record<string, unknown>;
    assert.equal(error.code, "INVALID_JSON");
  });
});

// ---------------------------------------------------------------------------
// Tests: invoke() (remote capability invocation via sidecar)
// ---------------------------------------------------------------------------

describe("HiveAgent.invoke()", () => {
  let mockSidecar: http.Server;
  let sidecarPort: number;

  before(async () => {
    // Start a mock sidecar that responds to invoke-remote requests.
    mockSidecar = http.createServer((req, res) => {
      const chunks: Buffer[] = [];
      req.on("data", (c: Buffer) => chunks.push(c));
      req.on("end", () => {
        const body = Buffer.concat(chunks).toString("utf-8");
        const parsed = JSON.parse(body) as Record<string, unknown>;
        const url = req.url ?? "";

        if (url.includes("/capabilities/echo/invoke-remote")) {
          const inputs = parsed.inputs as Record<string, unknown>;
          const response = JSON.stringify({
            status: "success",
            outputs: { echoed: inputs.text },
            duration_ms: 10,
          });
          res.writeHead(200, {
            "Content-Type": "application/json",
            "Content-Length": Buffer.byteLength(response),
          });
          res.end(response);
        } else if (url.includes("/capabilities/fail-cap/invoke-remote")) {
          const response = JSON.stringify({
            status: "error",
            error: {
              code: "AGENT_OFFLINE",
              message: "target agent not available",
              retryable: true,
            },
          });
          res.writeHead(502, {
            "Content-Type": "application/json",
            "Content-Length": Buffer.byteLength(response),
          });
          res.end(response);
        } else {
          res.writeHead(404);
          res.end();
        }
      });
    });

    await new Promise<void>((resolve) => {
      mockSidecar.listen(0, "127.0.0.1", () => {
        const addr = mockSidecar.address();
        if (addr && typeof addr !== "string") {
          sidecarPort = addr.port;
        }
        resolve();
      });
    });
  });

  after(() => {
    mockSidecar.close();
  });

  it("should invoke a remote capability successfully", async () => {
    const agent = new HiveAgent({
      callbackPort: 0,
      sidecarUrl: `http://127.0.0.1:${sidecarPort}`,
    });

    const result = await agent.invoke("agent-b", "echo", { text: "hello" });
    assert.equal(result.echoed, "hello");
  });

  it("should throw on remote invocation failure", async () => {
    const agent = new HiveAgent({
      callbackPort: 0,
      sidecarUrl: `http://127.0.0.1:${sidecarPort}`,
    });

    await assert.rejects(
      () => agent.invoke("agent-b", "fail-cap", {}),
      (err: Error) => {
        assert.match(err.message, /target agent not available|HTTP 502/);
        return true;
      }
    );
  });

  it("should pass timeout parameter to sidecar", async () => {
    // We verify it doesn't error; the mock sidecar doesn't validate timeout
    // but the request format is correct.
    const agent = new HiveAgent({
      callbackPort: 0,
      sidecarUrl: `http://127.0.0.1:${sidecarPort}`,
    });

    const result = await agent.invoke("agent-b", "echo", { text: "hi" }, "5s");
    assert.equal(result.echoed, "hi");
  });
});

// ---------------------------------------------------------------------------
// Tests: Multiple agents on different ports
// ---------------------------------------------------------------------------

describe("Multiple HiveAgent instances", () => {
  let agent1: HiveAgent;
  let agent2: HiveAgent;
  const port1 = 19200 + Math.floor(Math.random() * 1000) + 1000;
  const port2 = port1 + 1;

  before(async () => {
    agent1 = new HiveAgent({ callbackPort: port1, agentId: "agent-1" });
    agent1.capability("ping", async () => ({ from: "agent-1" }));
    agent1.run();

    agent2 = new HiveAgent({ callbackPort: port2, agentId: "agent-2" });
    agent2.capability("ping", async () => ({ from: "agent-2" }));
    agent2.run();

    await Promise.all([waitForServer(port1), waitForServer(port2)]);
  });

  after(() => {
    agent1.stop();
    agent2.stop();
  });

  it("both agents respond independently", async () => {
    const [res1, res2] = await Promise.all([
      request("POST", `http://127.0.0.1:${port1}/handle/ping`, { inputs: {} }),
      request("POST", `http://127.0.0.1:${port2}/handle/ping`, { inputs: {} }),
    ]);

    assert.equal(res1.status, 200);
    const out1 = res1.body.outputs as Record<string, unknown>;
    assert.equal(out1.from, "agent-1");

    assert.equal(res2.status, 200);
    const out2 = res2.body.outputs as Record<string, unknown>;
    assert.equal(out2.from, "agent-2");
  });
});

File: sdk/go/hive/agent_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 The Hive Authors

package hive

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

// ---------------------------------------------------------------------------
// NewAgent / environment variable reading
// ---------------------------------------------------------------------------

func TestNewAgent_DefaultPort(t *testing.T) {
	// Clear any existing env vars that could interfere.
	t.Setenv("HIVE_CALLBACK_PORT", "")
	t.Setenv("HIVE_AGENT_ID", "")
	t.Setenv("HIVE_TEAM_ID", "")
	t.Setenv("HIVE_SIDECAR_URL", "")
	t.Setenv("HIVE_WORKSPACE", "")

	a := NewAgent()
	if a.port != "9200" {
		t.Errorf("expected default port 9200, got %s", a.port)
	}
}

func TestNewAgent_EnvVars(t *testing.T) {
	t.Setenv("HIVE_CALLBACK_PORT", "8888")
	t.Setenv("HIVE_AGENT_ID", "test-agent")
	t.Setenv("HIVE_TEAM_ID", "test-team")
	t.Setenv("HIVE_SIDECAR_URL", "http://127.0.0.1:9100")
	t.Setenv("HIVE_WORKSPACE", "/tmp/workspace")

	a := NewAgent()
	if a.port != "8888" {
		t.Errorf("expected port 8888, got %s", a.port)
	}
	if a.AgentID() != "test-agent" {
		t.Errorf("expected agent_id test-agent, got %s", a.AgentID())
	}
	if a.TeamID() != "test-team" {
		t.Errorf("expected team_id test-team, got %s", a.TeamID())
	}
	if a.Workspace() != "/tmp/workspace" {
		t.Errorf("expected workspace /tmp/workspace, got %s", a.Workspace())
	}
}

// ---------------------------------------------------------------------------
// Capability registration
// ---------------------------------------------------------------------------

func TestHandleCapability_Registration(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("greet", func(inputs map[string]any) (map[string]any, error) {
		return map[string]any{"msg": "hello"}, nil
	})

	a.mu.RLock()
	_, ok := a.capabilities["greet"]
	a.mu.RUnlock()
	if !ok {
		t.Fatal("expected capability 'greet' to be registered")
	}
}

func TestHandleCapability_PanicsOnEmptyName(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected panic for empty capability name")
		}
	}()

	a := newTestAgent()
	a.HandleCapability("", func(inputs map[string]any) (map[string]any, error) {
		return nil, nil
	})
}

func TestHandleCapability_PanicsOnNilHandler(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected panic for nil handler")
		}
	}()

	a := newTestAgent()
	a.HandleCapability("test", nil)
}

// ---------------------------------------------------------------------------
// Health endpoint
// ---------------------------------------------------------------------------

func TestHealthEndpoint(t *testing.T) {
	a := newTestAgent()
	mux := testMux(a)

	req := httptest.NewRequest(http.MethodGet, "/health", nil)
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusOK {
		t.Errorf("expected status 200, got %d", w.Code)
	}

	var resp map[string]string
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("invalid JSON: %v", err)
	}
	if resp["status"] != "healthy" {
		t.Errorf("expected status healthy, got %s", resp["status"])
	}
}

func TestHealthEndpoint_MethodNotAllowed(t *testing.T) {
	a := newTestAgent()
	mux := testMux(a)

	req := httptest.NewRequest(http.MethodPost, "/health", nil)
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusMethodNotAllowed {
		t.Errorf("expected status 405, got %d", w.Code)
	}
}

// ---------------------------------------------------------------------------
// Capability invocation (sidecar -> agent HTTP)
// ---------------------------------------------------------------------------

func TestHandleCapability_Success(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("greet", func(inputs map[string]any) (map[string]any, error) {
		name, _ := inputs["name"].(string)
		return map[string]any{"message": "Hello, " + name + "!"}, nil
	})
	mux := testMux(a)

	body := `{"inputs": {"name": "World"}}`
	req := httptest.NewRequest(http.MethodPost, "/handle/greet", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected status 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp map[string]any
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("invalid JSON: %v", err)
	}
	outputs, ok := resp["outputs"].(map[string]any)
	if !ok {
		t.Fatal("expected outputs in response")
	}
	if outputs["message"] != "Hello, World!" {
		t.Errorf("expected 'Hello, World!', got %v", outputs["message"])
	}
}

func TestHandleCapability_EmptyInputs(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("ping", func(inputs map[string]any) (map[string]any, error) {
		return map[string]any{"pong": true}, nil
	})
	mux := testMux(a)

	// Send with empty body.
	req := httptest.NewRequest(http.MethodPost, "/handle/ping", nil)
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected status 200, got %d: %s", w.Code, w.Body.String())
	}
}

func TestHandleCapability_HandlerError(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("fail", func(inputs map[string]any) (map[string]any, error) {
		return nil, fmt.Errorf("something went wrong")
	})
	mux := testMux(a)

	body := `{"inputs": {}}`
	req := httptest.NewRequest(http.MethodPost, "/handle/fail", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusInternalServerError {
		t.Fatalf("expected status 500, got %d", w.Code)
	}

	var resp map[string]any
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("invalid JSON: %v", err)
	}
	errObj, ok := resp["error"].(map[string]any)
	if !ok {
		t.Fatal("expected error object in response")
	}
	if errObj["code"] != "CAPABILITY_FAILED" {
		t.Errorf("expected code CAPABILITY_FAILED, got %v", errObj["code"])
	}
	if errObj["message"] != "something went wrong" {
		t.Errorf("expected message 'something went wrong', got %v", errObj["message"])
	}
}

func TestHandleCapability_NotFound(t *testing.T) {
	a := newTestAgent()
	mux := testMux(a)

	body := `{"inputs": {}}`
	req := httptest.NewRequest(http.MethodPost, "/handle/nonexistent", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusNotFound {
		t.Fatalf("expected status 404, got %d", w.Code)
	}

	var resp map[string]any
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("invalid JSON: %v", err)
	}
	errObj, ok := resp["error"].(map[string]any)
	if !ok {
		t.Fatal("expected error object in response")
	}
	if errObj["code"] != "CAPABILITY_NOT_FOUND" {
		t.Errorf("expected code CAPABILITY_NOT_FOUND, got %v", errObj["code"])
	}
}

func TestHandleCapability_InvalidJSON(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("test", func(inputs map[string]any) (map[string]any, error) {
		return map[string]any{}, nil
	})
	mux := testMux(a)

	req := httptest.NewRequest(http.MethodPost, "/handle/test", strings.NewReader("not json"))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusBadRequest {
		t.Fatalf("expected status 400, got %d", w.Code)
	}
}

func TestHandleCapability_MethodNotAllowed(t *testing.T) {
	a := newTestAgent()
	a.HandleCapability("test", func(inputs map[string]any) (map[string]any, error) {
		return map[string]any{}, nil
	})
	mux := testMux(a)

	req := httptest.NewRequest(http.MethodGet, "/handle/test", nil)
	w := httptest.NewRecorder()
	mux.ServeHTTP(w, req)

	if w.Code != http.StatusMethodNotAllowed {
		t.Fatalf("expected status 405, got %d", w.Code)
	}
}

// ---------------------------------------------------------------------------
// Invoke (agent -> sidecar -> remote agent)
// ---------------------------------------------------------------------------

func TestInvoke_Success(t *testing.T) {
	// Set up a mock sidecar server.
	sidecar := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			t.Errorf("expected POST, got %s", r.Method)
		}
		expectedPath := "/capabilities/greet/invoke-remote"
		if r.URL.Path != expectedPath {
			t.Errorf("expected path %s, got %s", expectedPath, r.URL.Path)
		}

		var req invokeRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatalf("decode request: %v", err)
		}
		if req.Target != "agent-b" {
			t.Errorf("expected target agent-b, got %s", req.Target)
		}

		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"outputs": map[string]any{"message": "Hello!"},
		})
	}))
	defer sidecar.Close()

	a := newTestAgent()
	a.sidecarURL = sidecar.URL

	outputs, err := a.Invoke(context.Background(), "agent-b", "greet", map[string]any{"name": "World"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if outputs["message"] != "Hello!" {
		t.Errorf("expected 'Hello!', got %v", outputs["message"])
	}
}

func TestInvoke_RemoteError(t *testing.T) {
	sidecar := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusInternalServerError)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]string{
				"code":    "CAPABILITY_FAILED",
				"message": "handler crashed",
			},
		})
	}))
	defer sidecar.Close()

	a := newTestAgent()
	a.sidecarURL = sidecar.URL

	_, err := a.Invoke(context.Background(), "agent-b", "greet", nil)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "CAPABILITY_FAILED") {
		t.Errorf("expected error to contain CAPABILITY_FAILED, got: %v", err)
	}
}

func TestInvoke_NoSidecarURL(t *testing.T) {
	a := newTestAgent()
	a.sidecarURL = ""

	_, err := a.Invoke(context.Background(), "agent-b", "greet", nil)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "HIVE_SIDECAR_URL not set") {
		t.Errorf("expected HIVE_SIDECAR_URL error, got: %v", err)
	}
}

func TestInvoke_EmptyTarget(t *testing.T) {
	a := newTestAgent()
	a.sidecarURL = "http://127.0.0.1:9100"

	_, err := a.Invoke(context.Background(), "", "greet", nil)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "target agent ID must not be empty") {
		t.Errorf("expected target error, got: %v", err)
	}
}

func TestInvoke_EmptyCapability(t *testing.T) {
	a := newTestAgent()
	a.sidecarURL = "http://127.0.0.1:9100"

	_, err := a.Invoke(context.Background(), "agent-b", "", nil)
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "capability name must not be empty") {
		t.Errorf("expected capability error, got: %v", err)
	}
}

func TestInvoke_ContextCancellation(t *testing.T) {
	// Use a listener that we never accept on — the client will connect but
	// the request will hang until the context deadline fires.
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()

	a := newTestAgent()
	a.sidecarURL = "http://" + ln.Addr().String()
	// Use a client without its own timeout so only the context controls it.
	a.httpClient = &http.Client{}

	ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
	defer cancel()

	_, invokeErr := a.Invoke(ctx, "agent-b", "greet", nil)
	if invokeErr == nil {
		t.Fatal("expected error from cancelled context")
	}
}

// ---------------------------------------------------------------------------
// Run / graceful shutdown
// ---------------------------------------------------------------------------

func TestRun_GracefulShutdown(t *testing.T) {
	a := newTestAgentWithPort(t)
	a.HandleCapability("test", func(inputs map[string]any) (map[string]any, error) {
		return map[string]any{"ok": true}, nil
	})

	ctx, cancel := context.WithCancel(context.Background())
	errCh := make(chan error, 1)
	go func() {
		errCh <- a.Run(ctx)
	}()

	// Give the server time to start.
	time.Sleep(100 * time.Millisecond)

	// Verify the server is responding.
	resp, err := http.Get("http://127.0.0.1:" + a.port + "/health")
	if err != nil {
		t.Fatalf("health check failed: %v", err)
	}
	resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}

	// Cancel context to trigger shutdown.
	cancel()

	select {
	case err := <-errCh:
		if err != nil {
			t.Fatalf("Run returned error: %v", err)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("Run did not return within 5 seconds")
	}
}

func TestRun_ServesCapabilities(t *testing.T) {
	a := newTestAgentWithPort(t)
	a.HandleCapability("echo", func(inputs map[string]any) (map[string]any, error) {
		return inputs, nil
	})

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	errCh := make(chan error, 1)
	go func() {
		errCh <- a.Run(ctx)
	}()

	// Give the server time to start.
	time.Sleep(100 * time.Millisecond)

	// Invoke the capability via HTTP.
	body := strings.NewReader(`{"inputs": {"msg": "hello"}}`)
	resp, err := http.Post("http://127.0.0.1:"+a.port+"/handle/echo", "application/json", body)
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}

	var result map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		t.Fatalf("decode response: %v", err)
	}
	outputs, ok := result["outputs"].(map[string]any)
	if !ok {
		t.Fatal("expected outputs in response")
	}
	if outputs["msg"] != "hello" {
		t.Errorf("expected 'hello', got %v", outputs["msg"])
	}

	cancel()
	select {
	case err := <-errCh:
		if err != nil {
			t.Fatalf("Run returned error: %v", err)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("Run did not return within 5 seconds")
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// newTestAgent creates an Agent with test defaults (no env dependency).
func newTestAgent() *Agent {
	return newAgentFromConfig(config{
		agentID:    "test-agent",
		teamID:     "test-team",
		sidecarURL: "http://127.0.0.1:9100",
		workspace:  "/tmp/test-workspace",
		port:       "9200",
	})
}

// newTestAgentWithPort creates an Agent bound to a random available port.
// The port is allocated by briefly listening, then immediately closed so
// the agent can bind to it.
func newTestAgentWithPort(t *testing.T) *Agent {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("allocate test port: %v", err)
	}
	port := fmt.Sprintf("%d", ln.Addr().(*net.TCPAddr).Port)
	ln.Close()

	return newAgentFromConfig(config{
		agentID:    "test-agent",
		teamID:     "test-team",
		sidecarURL: "http://127.0.0.1:9100",
		workspace:  "/tmp/test-workspace",
		port:       port,
	})
}

// testMux creates an http.ServeMux wired to the agent's handlers (for
// httptest usage without starting a real server).
func testMux(a *Agent) *http.ServeMux {
	mux := http.NewServeMux()
	mux.HandleFunc("/health", a.handleHealth)
	mux.HandleFunc("/handle/", a.handleCapability)
	return mux
}
